README.md
## Description

Deis (pronounced DAY-iss) is an open source PaaS that makes it easy to deploy and manage
applications on your own servers. Deis builds on [Kubernetes](http://kubernetes.io/) to provide
a lightweight, [Heroku-inspired](http://heroku.com) workflow.

![Deis Graphic](https://s3-us-west-2.amazonaws.com/get-deis/deis-graphic-small.png)

## About
This is an centos7 based image for running [fluentd](http://fluentd.org). It is built for the purpose of running on a kubernetes cluster.

This work is based on the [docker-fluentd](https://github.com/fabric8io/docker-fluentd) and [docker-fluentd-kubernetes](https://github.com/fabric8io/docker-fluentd-kubernetes) images by the fabric8 team. This image is in with [deis](https://github.com/deis/deis) v2 to send all log data to the [logger](https://github.com/deis/logger) component.

## Plugins

### [fluent-plugin-kubernetes_metadata_filter](https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter)
This plugin is used to decorate all log entries with kubernetes metadata.

### [fluent-plugin-elasticsearch](https://github.com/uken/fluent-plugin-elasticsearch)
Allows fluentd to send log data to an elastic search cluster. You must specify an `ELASTICSEARCH_HOST` environment variable for this plugin to work.

### [fluent-plugin-remote_syslog](https://github.com/dlackty/fluent-plugin-remote_syslog)
This plugin allows `fluentd` to send data to a remote syslog endpoint like [papertrail](http://papertrailapp.com). You can configure `fluentd` to talk to multiple remote syslog endpoints by using the following scheme:
* `SYSLOG_HOST_1=some.host`
* `SYSLOG_PORT_1=514`
* `SYSLOG_HOST_2=some.other.host`
* `SYSLOG_PORT_2=52232`

You can also set `SYSLOG_HOST` and `SYSLOG_PORT`..

### Deis Output
Deis output is a custom fluentd plugin that was written to forward data directly to deis components while filtering out data that we did not care about. We have 2 pieces of information we care about currently.

1) Logs from applications that are written to stdout within the container and the controller logs that represent actions against those applications. These logs are sent to an internal messaging system ([NSQ](http://nsq.io)) on a configurable topic. The logger component then reads those messages and stores the data in an ring buffer.

2) Metric data from the nginx based router. We take the log and parse out `request_time`, `response_time`, and `bytes_sent`. Each one of these metrics makes up a series that we will ultimately send to our InfluxDB system. Attached to each series is the host the data came from (where router is running) and the status code for that request.

The topics these messages are put on are configurable via environment variables.
* `NSQ_LOG_TOPIC`
* `NSQ_METRIC_TOPIC`


CHANGELOG.md
### v2.0.0-rc2 -> v2.0.0

#### Documentation

- [`e945cb5`](https://github.com/deis/fluentd/commit/e945cb57f79fab482c5b835599af2bcc9da3cb51) CHANGELOG.md: add entry for v2.0.0-rc2

### v2.0.0-rc1 -> v2.0.0-rc2

### v2.0.0-beta4 -> v2.0.0-rc1

#### Maintenance

 - [`82ed849`](https://github.com/deis/fluentd/commit/82ed849bc156d0b1cafd06c65a5e0b454b646010) Dockerfile: Refactor image to use ubuntu-slim
 - [`701c9cc`](https://github.com/deis/fluentd/commit/701c9ccf5b7b6edb8e0f602500bfefd2355fe181) syslog-plugin: Remove dead code

### v2.0.0-beta3 -> v2.0.0-beta4

#### Documentation

docs(CHANGELOG.md): update for v2.0.0-beta3
 - [`b941f64`](https://github.com/deis/fluentd/commit/b941f64b1e251456ad933b8a958519873c51af69) CHANGELOG.md: update for v2.0.0-beta3

#### Maintenance

- [`f73e095`](https://github.com/deis/fluentd/commit/f73e095aec3278154cd076e14ed6909c238812ff): fix: Create new UDPSender for each message we get
    The plugin was trying to reuse connection information but this meant that if the other end died the packets would never be sent. Instead we now keep a global list of host:ports we want to send data to and iterate that list creating a new sender each time. This is wrapped in a begin/rescue/ensure block for protection

### v2.0.0-beta2 -> v2.0.0-beta3

#### Fixes

 - [`30f4eef`](https://github.com/deis/fluentd/commit/30f4eef1c5ec406e247574f682ae556e07fe3016) conf: Capture all kubernetes log streams and cleanup generated conf

#### Maintenance

 - [`dd1f83f`](https://github.com/deis/fluentd/commit/dd1f83f0e0c38203322ec80c7297f711582e348a) travis.yml: Remove travis.yml

### v2.0.0-beta1 -> v2.0.0-beta2

#### Maintenance

 - [`c2910f1`](https://github.com/deis/fluentd/commit/c2910f1c89da495aa5123e067802d5932b41f4e8) makefile: Update makefile to push canary tag when building new image


rootfs/opt/fluentd/deis-output/test/test_out_deis.rb
require 'fluent/test'
require 'fluent/plugin/out_deis'

class DeisOutputTest < Test::Unit::TestCase
  CONFIG = %[tag foo.bar]

  def setup
    Fluent::Test.setup
    @valid_record = { "kubernetes" => { "container_name" => "deis-controller" } }
    @invalid_record = { }
    @valid_app_record = { "kubernetes" => { "labels" => { "app" => "foo", "heritage" => "deis" } } }
    @invalid_app_record = { "kubernetes" => { "labels" => { "foo" => "foo" } } }

    @valid_router_log = { "log"=>"[2016-05-31T16:56:12+00:00] - foo - 10.2.1.1 - - - 200 - \"GET / HTTP/1.0\" - 211 - - - \"ApacheBench/2.3\" - \"~^foo\\x5C.(?<domain>.+)$\" - 10.1.1.4:80 - foo.bar.io - 0.002 - 0.046\n" }

    @deis_output = Fluent::DeisOutput.new
  end

  def test_kubernetes_should_return_true_with_valid_key
    output = Fluent::DeisOutput.new
    assert_true(@deis_output.kubernetes?(@valid_record))
  end

  def test_kubernetes_should_return_false_with_invalid_key
    assert_false(@deis_output.kubernetes?(@invalid_record))
  end

  def test_from_container_should_return_true_with_valid_container_name
    assert_true(@deis_output.from_container?(@valid_record, "deis-controller"))
  end

  def test_from_container_should_return_false_with_invalid_container_name
    assert_false(@deis_output.from_container?(@valid_record, "deis-foo"))
  end

  def test_from_container_should_return_false_with_invalid_record
    assert_false(@deis_output.from_container?(@invalid_record, "deis-controller"))
  end

  def test_deis_deployed_app_should_return_true_with_valid_application_message
    assert_true(@deis_output.deis_deployed_app?(@valid_app_record))
  end

  def test_deis_deployed_app_should_return_false_with_valid_application_message
    assert_false(@deis_output.deis_deployed_app?(@invalid_app_record))
  end

  def test_deis_parse_router_message_should_return_valid_map
    router_metrics = @deis_output.parse_router_log(@valid_router_log["log"], "some.new.host")
    assert_true(router_metrics["app"] == "foo")
    assert_true(router_metrics["status_code"] == "200")
    assert_true(router_metrics["bytes_sent"] == 211.0)
    assert_true(router_metrics["response_time"] == 0.002)
    assert_true(router_metrics["request_time"] == 0.046)
    assert_true(router_metrics["host"] == "some.new.host")
  end
end


out_remote_syslog.rb
require "fluent/mixin/config_placeholders"
require "fluent/mixin/plaintextformatter"
require 'fluent/mixin/rewrite_tag_name'

module Fluent
  class RemoteSyslogOutput < Fluent::Output
    Fluent::Plugin.register_output("remote_syslog", self)

    config_param :hostname, :string, :default => ""

    include Fluent::Mixin::PlainTextFormatter
    include Fluent::Mixin::ConfigPlaceholders
    include Fluent::HandleTagNameMixin
    include Fluent::Mixin::RewriteTagName

    config_param :host, :string
    config_param :port, :integer, :default => 25

    config_param :facility, :string, :default => "user"
    config_param :severity, :string, :default => "notice"
    config_param :tag, :string, :default => "fluentd"

    def initialize
      super
      require "remote_syslog_logger"
    end

    def start
      super
    end

    def shutdown
      super
    end

    def emit(tag, es, chain)
      chain.next
      es.each do |time, record|
        record.each_pair do |k, v|
          if v.is_a?(String)
            v.force_encoding("utf-8")
          end
        end

        tag = rewrite_tag!(tag.dup)
        sender = RemoteSyslogLogger::UdpSender.new(@host, @port, {facility: @facility, severity: @severity, program: tag, local_hostname: @hostname})
        begin
          sender.transmit format(tag, time, record)
        rescue Exception => e
          puts "Error:#{e.message}"
        ensure
          sender.close if sender
        end
      end
    end
  end
end


_scripts/generate-changelog.sh
#!/usr/bin/env bash
# vim: et tw=0 sw=4 ts=4

REPOROOT=${REPOROOT:=https://github.com/deis/fluentd}
TYPES=${TYPES:=feat(;Features fix(;Fixes docs(;Documentation chore(;Maintenance}

usage() {
    echo "Usage: $0 <from> [to]
The REPOROOT and TYPES variables may be used to customize what links are
generated and what is scraped from the commit logs. Current settings are:
    REPOROOT=$REPOROOT
    TYPES=$TYPES
"
}

# Print the information scraped from git
retrieve() {
    while read -r commit hash message; do
        # Extract the subsystem where type(subsystem): message
        SUBSYSTEM=$(echo "$message" | cut -d'(' -f2 | cut -d')' -f1 | sed 's/\*/\(all\)/g')
        # Extract the message where type(subsystem): message
        MESSAGE=$(echo "$message" | awk -F ")" '{ print $2}' | sed 's/://' | cut -f2- -d' ')
        # Generate a link to the full legal commit on GitHub
        LINK="$REPOROOT/commit/$hash"
        # Echo all this in a way that makes the commit hash and message a link
        # to the commit in markdown
        echo ' -' "[\`$commit\`]($LINK)" "$SUBSYSTEM": "$MESSAGE"
    done < <(git --no-pager log --oneline --merges --oneline --format="%h %H %b" --grep="$1" "$FROM".."$TO")
    # Scrape the information from git
}

# Wrap feature type and show its relevant commits
subheading() {
    echo "#### $1"
    echo
    retrieve "$2"
    echo
}

main() {
    # Print usage summary if user didn't specify a beginning
    if [ -z "$1" ]; then
        usage
        exit 1
    fi

    FROM=$1
    TO=${2:-"HEAD"}

    printf "### %s -> %s\n\n" "$FROM" "$TO"

    # Iterate over the types of messages specified
    for LEGALTYPE in $TYPES; do
        SHORT=$(echo "$LEGALTYPE" | cut -f1 -d';')
        LONG=$(echo "$LEGALTYPE" | cut -f2 -d';')
        subheading "$LONG" "$SHORT"
    done
}

if [ $SHLVL -eq 2 ]; then
    # If this is being run as a command
    main "$@"
else
    # Otherwise this is being sourced
    unset -f main
    unset -f usage
fi


rootfs/opt/fluentd/deis-output/lib/fluent/plugin/out_deis.rb
require 'fluent/mixin/config_placeholders'
require 'fluent/mixin/plaintextformatter'
require 'fluent/mixin/rewrite_tag_name'
require 'fluent/mixin/deis'
require 'fluent/output'

module Fluent
  class DeisOutput < Output
    Fluent::Plugin.register_output("deis", self)

    include Fluent::Mixin::PlainTextFormatter
    include Fluent::Mixin::ConfigPlaceholders
    include Fluent::HandleTagNameMixin
    include Fluent::Mixin::RewriteTagName
    include Fluent::Mixin::Deis

    config_param :tag, :string, :default => ""
    config_set_default :output_include_time, false
    config_set_default :output_include_tag, false
    config_set_default :num_threads, 5
    config_set_default :flush_thread_count, 5

    def initialize
      super
      @logger_nsq = nil
      @influx_nsq = nil
      @log_topic = ENV['NSQ_LOG_TOPIC'] || "logs"
      @metric_topic = ENV['NSQ_METRIC_TOPIC'] || "metrics"
    end

    def start
      super
    end

    def shutdown
      super
      @logger_nsq.terminate if @logger_nsq
      @influx_nsq.terminate if @influx_nsq
    end

    def emit(tag, es, chain)
      es.each do |time, record|
        if from_controller?(record) || deis_deployed_app?(record)
          @logger_nsq ||= get_nsq_producer(@log_topic)
          push(@logger_nsq, record) if @logger_nsq
        end

        if from_router?(record)
          @influx_nsq ||= get_nsq_producer(@metric_topic)
          begin
            data = build_series(record)
            if data
              line = data.map do |point|
                InfluxDB::PointValue.new(point).dump
              end.join("\n".freeze)
              push(@influx_nsq, line) if @influx_nsq
            end
          rescue Exception => e
            puts "Error:#{e.backtrace}"
          end
        end
      end
      chain.next
    end
  end
end


rootfs/opt/fluentd/deis-output/lib/fluent/mixin/deis.rb
require 'json'
require 'nsq'
require 'influxdb'

module Fluent
  module Mixin
    module Deis
      LOGGER_URL = "http://#{ENV['DEIS_LOGGER_SERVICE_HOST']}:#{ENV['DEIS_LOGGER_SERVICE_PORT_HTTP']}/logs"
      INFLUX_HOST = "#{ENV['DEIS_MONITOR_INFLUXAPI_SERVICE_HOST']}"
      INFLUX_PORT = "#{ENV['DEIS_MONITOR_INFLUXAPI_SERVICE_PORT_TRANSPORT']}"
      INFLUX_DATABASE = ENV['INFLUX_DATABASE'] || "kubernetes"
      NSQ_URL = "#{ENV['DEIS_NSQD_SERVICE_HOST']}:#{ENV['DEIS_NSQD_SERVICE_PORT_TRANSPORT']}"

      def kubernetes?(message)
        return message["kubernetes"] != nil
      end

      def from_controller?(message)
        if from_container?(message, "deis-controller")
          return message["log"] =~ /^(INFO|WARN|DEBUG|ERROR)\s+(\[(\S+)\])+:(.*)/
        end
        return false
      end

      def from_router?(message)
        return from_container?(message, "deis-router")
      end

      def from_container?(message, regex)
        if kubernetes? message
          return true if Regexp.new(regex).match(message["kubernetes"]["container_name"]) != nil
        end
        return false
      end

      def deis_deployed_app?(message)
        if kubernetes? message
          labels = message["kubernetes"]["labels"]
          return true if message["kubernetes"]["namespace_name"] != "deis" && labels["heritage"] == "deis" && labels["app"] != nil
        end
        return false
      end

      def push(producer, value)
        begin
          if value.kind_of? Hash
            producer.write(value.to_json)
          else
            producer.write(value)
          end
        rescue Exception => e
          puts "Error:#{map.message}"
        end
      end

      def build_series(message)
        metric = parse_router_log(message["log"], message["kubernetes"]["host"])
        if metric
          tags = { app: metric["app"], status_code: metric["status_code"], host: metric["host"] }
          data = [
            {
              series: 'deis_router_request_time_ms',
              values: { value: metric["request_time"] },
              tags: tags
            },
            {
              series: 'deis_router_response_time_ms',
              values: { value: metric["response_time"] },
              tags: tags
            },
            {
              series: 'deis_router_bytes_sent',
              values: { value: metric["bytes_sent"] },
              tags: tags
            },
          ]
          return data
        end
        return nil
      end

      # {"log"=>"[2016-05-31T16:56:12+00:00] - foo - 10.164.1.1 - - - 200 - \"GET / HTTP/1.0\" - 211 - \"-\" - \"ApacheBench/2.3\" - \"~^foo\\x5C.(?<domain>.+)$\" - 10.167.243.4:80 - foo.devart.io - 0.002 - 0.046\n"}
      # request_time - request processing time in seconds with a milliseconds resolution; time elapsed between the first bytes were read from the client and the log write after the last bytes were sent to the client
      # response_time - keeps time spent on receiving the response from the upstream server; the time is kept in seconds with millisecond resolution.
      def parse_router_log(message, host)
        begin
          split_message = message.split(" - ")
          return nil if split_message.length < 14
          metric = {}
          metric["app"] = split_message[1].strip
          metric["status_code"] = split_message[4].strip
          metric["bytes_sent"] = split_message[6].strip.to_f
          metric["response_time"] = split_message[12].strip.to_f
          metric["request_time"] = split_message[13].strip.to_f
          metric["host"] = host
          return metric
        rescue Exception => e
          puts "Error:#{e.message}"
          return nil
        end
      end

      def debug?
        ENV["DEBUG"] == "true"
      end

      def get_nsq_producer(topic)
        begin
          puts "Creating nsq producer (#{NSQ_URL}) for topic:#{topic}"
          return Nsq::Producer.new(nsqd: NSQ_URL, topic: topic)
        rescue Exception => e
          puts "Error:#{e.message}"
          return nil
        end
      end
    end
  end
end
